### PR TITLE
Case search labels migration

### PR DESCRIPTION
**Pull request by mkangia**
## Summary

Review by commit

As a part of https://dimagi-dev.atlassian.net/browse/USH-901 we want to add multimedia support for "Search" and "Search Again" labels.
In exploration in https://github.com/dimagi/commcare-hq/pull/29714/files#r632185204 it was discussed to do a migration for the current labels first.
So, this PR migrates
1. command_label to search_label
2. again_label to search_again_label

I retained the original fields to write test for the migration, assuming that I can't populate an app doc with removed properties.
Open to suggestions to
1. ~~remove the fields and the test added, as the last commit after code freeze on this, to run tests without the properties once. Removing the test as well because it won't work after removing the properties.~~
2. remove it in follow up PR instead
3. ~~update test to still work without properties in model~~

I will be requesting a QA post one approval on PR

The approach to roll this out was revisited after initial plan which was to let it go like a normal migration but then finalized with https://github.com/dimagi/commcare-hq/pull/29735#issuecomment-849519688

## Feature Flag

`USH_CASE_CLAIM_UPDATES` and `SYNC_SEARCH_CASE_CLAIM`

## Product Description

The only visible user-facing change is that in the app translations download, the "Search" label will now be search_label instead of command_label

## Safety Assurance

- [x] Risk label is set correctly
- [ ] All migrations are backwards compatible and won't block deploy
- [x] The set of people pinged as reviewers is appropriate for the level of risk of the change
- [x] If QA is part of the safety story, the "Awaiting QA" label is used
- [x] I have confidence that this PR will not introduce a regression for the reasons below

### Automated test coverage

Added test for the migration itself. There are existing tests as well for other changes.

### QA Plan

https://dimagi-dev.atlassian.net/browse/QA-3044

### Safety story

I searched for all occurrences of "command_label" and "again_label" in the code, and then "commandlabel" and "againlabel", for javascript matches, to update them if needed.

Got a green test build. 

Did a local UI test as well and then request for QA for migration and case search feature sanity check, by asking them to setup a working case search instance and then deploy this to migrate before the final round of testing.
Overall the risk isn't major. The feature is also behind a feature flag.

### Rollback instructions

The PR has a migration. 

The migration was run on production and india in advance using setup_release to avoid having it run during deploy.

If this needs to be reverted, after a successful deploy, depending on how much time has passed since merge, we might need to add a reverse of what is happening in the migration command, which should be easy to add. The reverse can be added in the code, to revert on all environments or just run on an effected environment.
If it failed mid deploy, then fine to just leave it there.

Please note that the migration had trouble running in one go on staging, with couch unable to keep up with batch size, after which it was just re-run couple of times which made it proceed, and then in the last attempt the batch size of updates was also reduced to 10. So no issues were found with the migration itself but just infra struggling with fetching docs for big apps.
Ultimately, it was reduced down to 1 doc at a time, when even 10 was big for apps on india. So there should be no issue with the migration now.

### CODE REVIEW

File: corehq/apps/app_manager/models.py
Comment (mkangia) [inline corehq/apps/app_manager/models.py:2099]: open to suggestions if this should be text or anything else so that the code references read well `search_config.search_label.text` instead of `search_config.search_label.label`

File: corehq/apps/app_manager/management/commands/migrate_case_search_labels.py
Comment (mkangia) [inline corehq/apps/app_manager/management/commands/migrate_case_search_labels.py:20]: command_label would always be present with the default values for any app doc that has been saved, so just `if command_label` didn't seem useful here.
So thought of updating if they are different, which I believe would be always true as well, comparing default dict with empty. 
Not fully sure if the `if` here is even useful and we should just update app docs.
Open to thoughts on this if condition here.

Comment (mkangia) [reply corehq/apps/app_manager/management/commands/migrate_case_search_labels.py]: I realized later that `module['search_config']['search_label']['label']` would fail on old docs, so updated in a later commit to use `get`

**Comment by orangejenny** [comment]
I wasn't anticipating also renaming the attributes, but I suppose as long as you're doing a migration, it's a good time to do it, I do think the new labels are clearer. In any case, thank you for being thorough.

File: corehq/apps/app_manager/models.py
Comment (orangejenny) [reply corehq/apps/app_manager/models.py:2099]: `label` sounds good to me

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
> 2. remove it in follow up PR instead

This sounds fine to me, since you're doing a followup PR for multimedia anyway.

File: corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py
Comment (snopoke) [inline corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: would be nice if you could use `run_once_off_migration` here instead of doing it manually. That would require changing the management command to allow passing all the enabled domains at once instead of individually.

**Comment by mkangia** [comment]
@orangejenny 

> I do think the new labels are clearer.

Agreed. Each time I read `command_label`, I felt a disconnect. I am not sure why that was used as the name originally. But it definitely makes the migration easier now since the name of the property has changed. Would have felt bad, changing if it was already case_search_label to case_search_label_2 or something :)

**Comment by mkangia** [comment]
@orangejenny 

> This sounds fine to me, since you're doing a followup PR for multimedia anyway.

Cool. I was just thinking as a follow up post deploy as a standalone PR. But good either way.

File: corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py
Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: This would need changes to the base management command, `AppMigrationCommandBase`.

Are there advantages to using that here?

From the documentation
```
This will give the user directions for running the command
    manually if the command fails or has since been removed from the codebase
```

We probably can remove the management command once its deployed to all environments or in a certain future, or possibly get rid of the migration and the command both once this is deployed on all envs.

Comment (snopoke) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: The trouble is supporting this for 3rd part envs. We could indeed just delete this after some period of time though I don't know what that period is.

**Comment by mkangia** [comment]
Tested this locally, and things are good before/during/after the migration.
I could test that
1. all apps got the new properties added with the values from the currently set labels
2. any app that had set its own labels, after the code update but before the migration was run, in the new properties, was retained as it is. This scenario is possible only locally I believe

I could then update new labels post migration. Old ones remained as it is.

**Comment by mkangia** [comment]
I made an update to the migration from thoughts on https://github.com/dimagi/commcare-hq/pull/29735#discussion_r633697094 that seemed like the best fit, considering,
1. old labels would always be present on the doc
2. new labels would not be present if the migration is ran on app docs, that have not been saved since code changes

So just updating when old label is present but new one is not, seemed like the best fit.

File: corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py
Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: This is behind a feature flag `USH_CASE_CLAIM_UPDATES` if that makes any difference.
I am leaning towards keeping it as it is. Let me know if there is a blocking change @snopoke

**Comment by mkangia** [comment]
QA requested: https://dimagi-dev.atlassian.net/browse/QA-3044

File: corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py
Comment (orangejenny) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: To be particular, this isn't exactly behind `USH_CASE_CLAIM_UPDATES`, it'll also affect domains with `SYNC_SEARCH_CASE_CLAIM` which is a more widely-used flag, though I don't know if there are local hosters using it.

Regarding timeline if we later delete this, we recommend local hosters deploy every two weeks: http://dimagi.github.io/commcare-cloud/system/maintenance-expectations.html

Comment (snopoke) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: > Regarding timeline if we later delete this, we recommend local hosters deploy every two weeks: http://dimagi.github.io/commcare-cloud/system/maintenance-expectations.html

I think there's a difference between what we recommend for deploy and what we 'support' for keeping migrations working. FYI this is already on the agenda for the next cross-divisional meeting.

Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: @orangejenny thanks for pointing that out.

It is because of this I thought this was the USH feature flag. It's only visible on the UI if this feature flag is enabled.

https://github.com/dimagi/commcare-hq/blob/d58656b0262958a8c5f4c98718c9549201e6ff29/corehq/apps/app_manager/templates/app_manager/partials/modules/case_search_properties.html#L111-L132

Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: > I think there's a difference between what we recommend for deploy and what we 'support' for keeping migrations working. FYI this is already on the agenda for the next cross-divisional meeting

So if we remove the migration from code, the migrations otherwise would still work?

Would be good to finalize on this, there is at least one more management command `migrate_case_search_relevant` + its migration, added just before the one added in this PR, which we can remove as well, if it is considered okay to remove

Comment (orangejenny) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: @mkangia I see what you mean about the flags. For historical context (which I had forgotten until you pointed this out): the `command_label` attribute was part of the original case claim feature, but it wasn't originally visible in the UI. USH made the changes to add it to the UI, make it translate-able, and also added the "Search Again" workflow. So the labels are only visible in the UI if the the USH feature flag is on, but app docs can still have the `command_label` attribute even if they use the regular case claim flag not the USH flag. Plus, the USH flag was added some time after the USH changes were made.

Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: Woops thanks. I will update the PR description to include both feature flags.
can you confirm if there is anything that needs to be handled differently as far as this PR is concerned?

Also,
If it was not in the UI, how was it set? via backend? and will there will many domains that would have that set but not have the USH specific feature flag enabled?

Comment (orangejenny) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: The code in this PR looks fine regarding flags.

> If it was not in the UI, how was it set? via backend? and will there will many domains that would have that set but not have the USH specific feature flag enabled?

Prior to USH putting it in the UI, apps would have had the `command_label` attribute, but it would always just be set to the default value, they had no way to change it. Then USH put it in the UI, under the regular search & claim flag. Then a month or two later, USH moved the UI behind the USH flag. So the only domains that aren't using the USH flag and have a non-default value would be any that happened to see the UI and change the label before it was switched to the USH flag. I doubt that's many, if any, domains.

Comment (mkangia) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: sweet, thanks.

Comment (snopoke) [reply corehq/apps/app_manager/migrations/0018_migrate_case_search_labels.py:23]: Fine to leave the migration as is. Can always be changed at a later point.

**Comment by mkangia** [comment]
Ran this on staging today. It was a bit of a trouble with couch failing to keep up with apps being requested. It even went down in one of the attempts.
Eventually, after my change to chunk size it went through just fine.
The migration can be run repeatedly with no issues, in case of failures so not considering making any changes for it. Hopefully it should be smooth on production.

**Comment by orangejenny** [comment]
@mkangia I overwrote your changes to the app.json files (the 3 template files and manage-multimedia.json) to get a staging deploy out, due to conflicts with https://github.com/dimagi/commcare-hq/pull/29782 - can you reapply those changes?

**Comment by mkangia** [comment]
thanks for handling that conflict @orangejenny , updated for my changes in those files.

**Comment by mkangia** [comment]
Hey @orangejenny @snopoke 

QA is through and found no issues, https://dimagi-dev.atlassian.net/browse/QA-3044?focusedCommentId=145041

I updated the "Rollback instructions" with some more specifics. Thinking of leaving the "All migrations are backwards compatible and won't block deploy" unchecked because of that.

Let me know if there are any more concerns or else would then go ahead and merge this now. Will coordinate the deploy of this on prod, whenever we have next deploy, and flag it on #dev-support.

**Comment by orangejenny** [comment]
@mkangia Is it feasible for this to be deployed while you're definitely online? I think the migration has some risk of having couch problems, as when you ran into on staging, and it's tough for a deployer without good context on this PR to decide when to cancel the migration if they run into issues.

An alternative could be to add more logging to the migration so that the deployer has an idea of how how far along the migration is.

**Comment by mkangia** [comment]
@orangejenny , yup! i will be online and coordinate this when it is deployed. 
I am assuming this would get deployed in IST, so should be easy to coordinate.

There is logging for at least the domains being traversed, so there is some idea of how far it has gone.
The good thing is that on re-attempt, the migration would effectively begin from where it left, because it won't save any docs before that point

**Comment by snopoke** [comment]
@mkangia I suppose you can't run this outside of a deploy because it has to go out concurrently with the model changes right?

**Comment by mkangia** [comment]
hey @snopoke 

Another way would be
1. add the properties on the document as placeholders, deploy
2. run the migration post deploy, to populate them
3. deploy this change, but also let the migration run again to update in case the old fields and new fields don't match, which would reduce the number of updates being done during deploy by 99%, Since i don't think people update these labels much.

I am hoping that reducing the chunk size done in the last commit would make the process smooth and not run into issues, since that was the point (loading the apps) where the errors were happening. So I am leaning towards not doing this differently.

**Comment by mkangia** [comment]
Hey @snopoke 

> What I do think you might want to do though is to run this from a limited release immediately before deploy. Doing it that way will allow you to handle any issues that come up much more easily than if they come up during the deploy

The migration really needs to be run just once, as of its current state where it populates the new fields if they are not present.
Correct me if i am wrong here, if we do a limited release and migrate, won't the new fields get removed if someone saves the application on the environment with the currently deployed code that does not have the new fields?
The issue I saw on staging was with fetching docs, so unless we are avoiding that completely during the deploy, the possibility of running in issues would still be there.

I am happy to run a dry test for the migration i.e run it without any saves, to get more confidence in the migration, if that helps.

> I see there's no way to 'resume' (no reiterate over apps from completed domains) if it fails. Considering the number of domains that have the FF enabled and I think it would make sense to try and accomplish that somehow. Perhaps a new command that accepts a CSV of 'completed' domains. The command could write the list of completed domains to a CSV as it goes.

That makes sense. It should be possible to add something for that if needed.

**Comment by snopoke** [comment]
> Hey @snopoke
> 
> > What I do think you might want to do though is to run this from a limited release immediately before deploy. Doing it that way will allow you to handle any issues that come up much more easily than if they come up during the deploy
> 
> The migration really needs to be run just once, as of its current state where it populates the new fields if they are not present.

I'm not suggesting we run it more than once. I'm suggesting this workflow:

1. Run migration manually via management command
2. Once complete, fake the Django migration to prevent it from running again during deploy
3. Deploy

> Correct me if i am wrong here, if we do a limited release and migrate, won't the new fields get removed if someone saves the application on the environment with the currently deployed code that does not have the new fields?

This is still a possibility when running it in a deploy. Until the deploy completes the site will be running old code which means that while the migration is running or before deploy completes any app saves could happen. That won't cause the new values to be removed but if someone changes one of the fields that are getting migrated after the specific app has been migrated then the new value will be lost.

> The issue I saw on staging was with fetching docs, so unless we are avoiding that completely during the deploy, the possibility of running in issues would still be there.
> 
> I am happy to run a dry test for the migration i.e run it without any saves, to get more confidence in the migration, if that helps.

I think that would be a good idea since it will also give you an idea of how long the migration will run for.

**Comment by mkangia** [comment]
> I think that would be a good idea since it will also give you an idea of how long the migration will run for.

Cool, i will run that and report back on findings to take a final call on how to roll this out.

> This is still a possibility when running it in a deploy.

Agreed. I think it makes sense to do a sanity check post migration + deploy anyway, to cross check if any app has different values for the two set of fields.

**Comment by mkangia** [comment]
Using this snippet to do a dry run

```
from corehq import toggles
from corehq.apps.app_manager.management.commands.helpers import get_all_app_ids
from corehq.apps.app_manager.models import Application

from corehq.util.couch import send_keys_to_couch

from dimagi.utils.chunked import chunked

apps_to_be_updated = {}

db = Application.get_db()

for domain in sorted(toggles.SYNC_SEARCH_CASE_CLAIM.get_enabled_domains()):
    app_ids = get_all_app_ids(domain=domain, include_builds=True)
    chunksize = 10
    for chunk in chunked(app_ids, chunksize):
        for res in send_keys_to_couch(db, keys=chunk):
            app_doc = res.get('doc')
            for module in app_doc.get('modules', []):
                if module.get('search_config'):
                    if domain in apps_to_be_updated:
                        apps_to_be_updated[domain] += 1
                    else:
                        apps_to_be_updated[domain] = 1
                    break  # break module iteration, move to next app
```

**Comment by mkangia** [comment]
Will run this tomorrow, 26th may, noon IST.

**Comment by mkangia** [comment]
On prod, 
Dry run took roughly 30 mins to run, total 34,849 applications would need to be migrated.
```
Start time (UTC) 2021-05-26 06:53:30.122523                                                                                         
End Time (UTC) 2021-05-26 07:22:57.875295
```
This was just loading the apps and not actually updating, so eventually this would take hours.
Thanks @snopoke for making me pursue this.

Definitely can't run this during deploy, so going with @snopoke suggestion
```
    Run migration manually via management command
    Once complete, fake the Django migration to prevent it from running again during deploy
    Deploy
```
Will then run a sanity check after the deploy

But overall since this can take very long to run, This should be run way before the deploy, at least 8-10 hours to be safe and not block deploy? that means it would need to be run outside off hours.
The load should not be high on couch, since we are fetching 10 apps at a time? 
I was observing 3 couch graphs, and only [one](https://app.datadoghq.com/dashboard/hr7-uza-kw5/couchdb-dashboard?from_ts=1621997817571&fullscreen_end_ts=1622014746000&fullscreen_paused=true&fullscreen_section=overview&fullscreen_start_ts=1622011146327&fullscreen_widget=16&live=true&to_ts=1622012217571) showed a slight jump during dry run.

Let me know your thoughts @snopoke @orangejenny

**Comment by snopoke** [comment]
@mkangia this poses a slight challenge. If you run it again after deploy how would you know if a different value is because the app changed before the deploy or after deploy? ie. how would you distinguish between these two scenarios?

1. value is edited after the migration but before deploy: copy the old value to the new location
2. value is edited after deploy: keep the new value

One method would be to keep track of the values at the time of migration and use them when re-running it afterwards.

**Comment by mkangia** [comment]
thanks for raising that @snopoke 

It is true that any sort of off deploy migration would need more steps.
I have added [this](https://docs.google.com/document/d/1CLt_xEk73mgGL6W_RITMFETosb8D3nJ7gePU3x1BOHk/edit#) document to consolidate rollout plan 

@snopoke  @orangejenny 
Seeing how challenging the rollout is, i want to propose an alternate that I was considering earlier but left it assuming roll out would be straight forward.

How about if we just migrate where the value for the old labels is something other than the default values?

Explanation for the same
1. The default value is same for both old and new labels
2. Anytime the application doc is wrapped, the default values will be present if the application doc has nothing set for new labels

So any time an app is loaded for use, it would load the default values for the new labels and use it, which would be the same as the old labels.
If it's saved, the value for new labels is persisted.
If its simply used to make build files or show in UI, the new labels would be present.
So its simply like when we add a new property to a couch doc and expect it to be present everywhere.

I believe this would work for saved releases as well.

I believe this should work but want a better opinion from you if this would do and we should consider it.
Tomorrow morning, then, I can run a fresh dry run that gets the number of apps with labels different from default values and i am hoping it would be a smaller number.

My original migration plan was based on how such a migration was done earlier in https://github.com/dimagi/commcare-hq/pull/29073/files#diff-9051d0b99d8eddae0e30d49c91152b6c524dfbffee513d963ad1a6bd29a7c123 which I guess needed all docs to be re-saved.

**Comment by orangejenny** [comment]
@mkangia If you only did current applications, would that be quick enough to happen during deploy? You could do saved builds before/after the deploy, since those don't change. Then the risk is reduced to revert workflows, which we might be able to mitigate another way.

**Comment by mkangia** [comment]
> @mkangia If you only did current applications, would that be quick enough to happen during deploy?

I believe so. I can confirm that when i do a dry run again tomorrow morning. 

> You could do saved builds before/after the deploy, since those don't change.

I think after deploy would be better to not have deploy blocked on anything.

> Then the risk is reduced to revert workflows, which we might be able to mitigate another way.

Can you tell more about this?

**Comment by orangejenny** [comment]
> Then the risk is reduced to revert workflows, which we might be able to mitigate another way.

If you migrate saved builds after deploy, the only risk is that someone reverts to an unmigrated old build, re-introducing the old attributes into their app. I think this risk is pretty slim, but you could mitigate it by updating [revert_to_copy](https://github.com/dimagi/commcare-hq/blob/012ba9749e0930ef5a293a8940ed844f71d679da/corehq/apps/app_manager/views/releases.py#L361) to do the migration on the saved build before reverting. That code could then be deleted once the migration is run across all saved builds.

**Comment by orangejenny** [comment]
> How about if we just migrate where the value for the old labels is something other than the default values?

Noting that @mkangia and I chatted about this approach in more detail and I think it's a reasonable one.

**Comment by mkangia** [comment]
Hey @orangejenny @snopoke 

Thanks for keeping up with me through this.

My dry run, this morning revealed interesting facts but nothing useful for this roll out.

So,
I took a step back, and instead of forcing this to work, took an alternate approach that would let me have my peace and not worry about things going wrong

So i have added,
https://github.com/dimagi/commcare-hq/pull/29810 which sets things up first.

Once that is merged, I will merge master here and remove the code to copy old labels to new ones.

Post next deploy, on prod, I would
- merge master branch here to bring this up to date
- Setup a private release using `mk/case_search_labels/migrate`
1. run the management command
2. fake the migration on prod by running `./manage migrate --fake app_manager 0018_migrate_case_search_labels`
3. merge this branch

I am hoping that after that, there would be no extra step needed for next deploy.

Let me know your thoughts. Feel free to comment on either PR.

**Comment by mkangia** [comment]
fyi @snopoke @orangejenny 

1. updated this PR with master, which removed changes done in https://github.com/dimagi/commcare-hq/pull/29810/commits/184aa7e658acd1f0c85ecd350db4608295e38f14 and keep it as they were in this PR (just use the new labels for bulk upload and UI updates) but retained the test which was good to have as a test for new labels since there was no test for labels on upload.
2. also removed an unwanted change i had pushed earlier accidentally https://github.com/dimagi/commcare-hq/pull/29735/commits/b6d829e289b1d09222c30abc8f96543fff9fa57f#diff-26cfc4ca190148f153c1caf38caf06d119e366771d148aba72520c6fc6f3819b

**Review by orangejenny — APPROVED** [review]

**Comment by orangejenny** [comment]
> fake the migration on prod by running ./manage migrate --fake app_manager 0018_migrate_case_search_labels

Will this work before the branch has been merged?

**Comment by mkangia** [comment]
> Will this work before the branch has been merged?

You mean in the release setup? Where i will be running this from?
It would be from this branch, so it will be present.
I believe It's suppose to just put an entry in the `django_migrations` table.

**Comment by mkangia** [comment]
Migration now completed on production.
Also faked the migration so that it is not run on deploy.

Will run it on india as well tonight, see a bunch of domains there as well, https://india.commcarehq.org/hq/flags/edit/search_claim/ but should not be a blocker for merge.
@snopoke @orangejenny can i get a final :+1: to merge this now?

**Comment by mkangia** [comment]
Left "All migrations are backwards compatible and won't block deploy" as unchecked because I think this cant be rolled back without repopulating the old values.

**Review by orangejenny — APPROVED** [review]

**Comment by mkangia** [comment]
I ran this on india as well. There were 46 domains but finished in roughly 10 mins or less.

However, it ran into issue fetching apps at once, due to some app being really big in size so i reduced the chunk_size to 1.

I can only say with some certainty that such a big app won't be present on local machines or other environments where the migration is run so i am going to update the chunk_size to be 1 itself.

**Review by orangejenny — APPROVED** [review]

**Pull request merged**
